### ISSUE TITLE
Support for no id when originating from client request

### ISSUE THREAD

**Issue by stevee2112** [neomerx/json-api#175]
There is an exception in the json-api spec regarding when a resource object originates from a client and represents a new resource.

"Exception: The id member is not required when the resource object originates at the client and represents a new resource to be created on the server."

http://jsonapi.org/format/#document-structure.

Currently it does not appear that this exception is supported and it may be a nice addition considering this is often a common use case for resource creation.

**Comment by neomerx** [comment]
@stevee2112 What kind of interface do you expect from the library? Suppose you've got a new resource and pass it to the encoder. How will it know that the resource was created with ID given and therefore ID field should be hidden?

**Comment by gregl83** [comment]
that's a good question...

i'd expect that a null value for an **id** property wouldn't default to an empty string which i believe is the current behavior.

**Comment by lindyhopchris** [comment]
hi! I actually think this is a misreading of the exception from the spec. My understanding is that the spec is saying when a client submits a new resource, it does not need to include the resource id (because it can only know it if using client-generated ids). However, the server should then send a `201` response, with the representation of the resource *including* the server generated id.

Therefore the spec says:

> If a POST request did not include a Client-Generated ID and the requested resource has been created successfully, the server MUST return a 201 Created status code... The response MUST also include a document that contains the primary resource created.

And that returned document would include the server generated id.

http://jsonapi.org/format/#crud-creating

**Comment by bravo-kernel** [comment]
I second @lindyhopchris. Also I don't see how library could help here, this has everything to do with posting data, not this library's concern is it? IMO the application should check for only two things when posting a new resource:

1. no `id` field: great, create the resource
2. `id` field not UUID: throw an error

> A server MAY accept a client-generated ID along with a request to create a resource. An ID MUST be specified with an id key, the value of which MUST be a universally unique identifier. 

http://jsonapi.org/format/#crud-creating

**Comment by neomerx** [comment]
As the spec points 'the resource object originates **at the client**'. That's not a typical scenario however the library is able to encode objects without ID. If the `Scheme` is dealing with newly created object without ID it can return `null` or empty string in `getId` method.
```php
    // in Schema
    public function getId($author)
    {
        return $author->isNew() ? null : $author->getId();
    }
```
As a result it will produce JSON such as below

```json
{  
   "data":{  
      "type":"people",
      "id":"",
      "attributes":{  
         "first_name":"Dan",
         "last_name":"Gebhardt"
      }
   }
}
```
The JSON contains `id` property which is in line with the spec 'The id member is not required when...'. Is it possible to remove it completely? Yes, it will require a small change in `\Neomerx\JsonApi\Document\Presenters\ElementPresenter::convertResourceToArray`

From
```php
        $representation = [
            Document::KEYWORD_TYPE => $resource->getType(),
            Document::KEYWORD_ID   => $resource->getId(),
        ];
```

To
```php
        $representation = [
            Document::KEYWORD_TYPE => $resource->getType(),
        ];
        if (($resourceId = $resource->getId()) !== null) {
            $representation[Document::KEYWORD_ID] = $resourceId;
        }
```
I'm not sure if the lib needs extra `if` that work only for a tiny fraction of use cases and do not improve compatibility with the spec. Your thoughts on this?

**Comment by lindyhopchris** [comment]
Ah I get what this is about. If you're using the package to serialize stuff as the client (i.e. if this package is being used as the client in the JSON API exchange), then yes you would expect the `id` member not to be included rather than being an empty string. But that's only if the resource object is the top-level member. I.e. if the resource object is in an array or is in the `included` member, it should always have an `id`.

However, if you're using the package as the server in the exchange, then actually you'd expect the `id` member to always be there when encoding - because regardless of whether it is a client or server generated id, it should always be there by the time you're encoding the resource.

(You can ignore my previous post... as I only use this package on the server-side, I had forgotten to consider that the package might be used on the client side instead.)

**Comment by stevee2112** [comment]
@neomerx, Your representation of this issue is spot on.  So is what you proposed as a change if this is deemed valuable.  I have two main comments on your most current response.

1) While I do agree this is a small use case, I still believe it actually does add some compatibility improvements with the spec, granted a small one.  As you pointed out, the majority (which is true for me) of use cases are based on a server response and not client orignation.  We are trying to use json-api across the board to design truly RESTful apis , and when posting data originating from the client it make sense that you would not have an id.  Obviously we can ignore the id if an empty string on the server side, but it pollutes the request object IMO and moves us away from what I believe was a concept supported by the original spec.

2) I would say there is a small caveat to simple removing the id if it is null.  I think there may be a small backwards compatibility issue with that change, in that, if for what ever reason someone is now expecting an empty response to indicate that no id is provided, and the key starts to disappearing after this change, it may cause issues for the users of the library.   That being said I suppose that is what versions are for.   Additionally, I suppose you could make this an encoder option, so people could opt in as needed when building a request object on the client side.  That would ensure that this issue could be pushed out as a patch update and truly have not backward incompatibility.

**Comment by gregl83** [comment]
@neomerx thanks for looking into it and publishing the lib. :+1: 

i do think the extra `if` is fair solution. using php7 return type hinting i'd expect `getId()` to become a nullable-string return type (`getId(): ?string`).

in my opinion, whatever the use-case may be, encoding null as an empty string can have unexpected results and makes it difficult to make sense of resource objects that have yet to be persisted.

this is a breaking change for some implementations.

**Comment by neomerx** [comment]
Just passed tests on `develop` branch and merged to `master`.

Adding `null` as a possible return value for `getId()` hasn't broken any tests. So I believe it should be fine as the return value fully depend on the developer anyway.

If no complaints till tomorrow it will be released as a next version.

**Comment by gregl83** [comment]
awesome, thanks again!

**Comment by stevee2112** [comment]
Agreed.  Thanks for the quick response!

**Comment by neomerx** [comment]
released in 1.0.1
